File: sources/kotlin/11_parallel_sum/main.kt
// task 11 parallel_sum — expected output: 7500000075000000
// build: kotlinc main.kt -include-runtime -d prog.jar    run: java -jar prog.jar    [native build: kotlinc-native -opt -o prog main.kt    native run: ./prog]
// note: this is the jvm row, using four java.lang.Thread workers; the native row has no java.lang and would need platform.posix pthreads instead.

private fun work(t: Long): Long {
    var acc = 0L
    var i = t * 25000000L
    val end = (t + 1L) * 25000000L
    while (i < end) {
        when (i % 4L) {
            0L -> acc += 1L
            1L -> acc += i
            2L -> acc += 2L * i
            else -> acc += 3L * i
        }
        i++
    }
    return acc
}

fun main() {
    val partials = LongArray(4)
    val workers = Array(4) { t ->
        Thread {
            partials[t] = work(t.toLong())
        }
    }
    for (w in workers) w.start()
    for (w in workers) w.join()
    var total = 0L
    for (p in partials) total += p
    println(total)
}

File: sources/python/11_parallel_sum/main.py
# task 11 parallel_sum — expected output: 7500000075000000
# build: python main.py | pypy main.py | graalpy main.py | nuitka --standalone main.py    run: python main.py
# note: threading.Thread is the standard-library mechanism, but the GIL serializes these four workers, so the sum is right and the speedup is not real (use multiprocessing for that).

import threading

partials = [0, 0, 0, 0]


def work(t):
    acc = 0
    for i in range(t * 25000000, (t + 1) * 25000000):
        match i % 4:
            case 0:
                acc += 1
            case 1:
                acc += i
            case 2:
                acc += 2 * i
            case 3:
                acc += 3 * i
    partials[t] = acc


threads = [threading.Thread(target=work, args=(t,)) for t in range(4)]
for th in threads:
    th.start()
for th in threads:
    th.join()

print(sum(partials))
